### ISSUE TITLE
Regression in Truffleruby-19.3.0 at memcpy() in C-ext

### ISSUE THREAD

**Issue by larskanis** [oracle/truffleruby#1822]
Truffleruby-19.3.0 fails with TypeError in two test cases of the pg gem, whereas Truffleruby-19.2.1 works:
```
TypeError:
  TruffleRuby doesn't have a case for the org.truffleruby.cext.CExtNodesFactory$StringPointerWriteNodeFactory$StringPointerWriteNodeGen node
  with values of type  String(com.oracle.truffle.object.DynamicObjectBasic) java.lang.Integer=0 java.lang.Long=8751735898823355977
```

It fails in [this line of the C code](https://github.com/ged/ruby-pg/blob/583403f9d617bd08a02bf750e8aa9f838677153c/ext/pg_text_encoder.c#L243). Full stack trace is here: https://travis-ci.org/larskanis/ruby-pg/jobs/615474667#L628

**Comment by eregon** [comment]
Thank you for the report.
@aardvark179 Could you look at this?

**Comment by aardvark179** [comment]
This is another problem with the new `memcpy` implementation in Sulong. Even if we ensure that the `RSTRING_PTR` is typed as a polyglot i8 array ``memcpy` still attempts to copy into it with i64s.

**Comment by aardvark179** [comment]
I have had a discussion with @rschatz and we have a plan to resolve this and #1772. This requires a small fix in TruffleRuby to declare the correct polyglot type for `RSTRING_PTR`, and some work in our LLVM interpreter to convert managed objects to native in certain cases. This will cause slightly more conversions to native objects than happens currently, but is relatively simple fix to resolve these issues until a more complete solution can be built.

**Comment by eregon** [comment]
We should try to solve this soon as it fails in pg's CI, for instance:
https://travis-ci.org/ged/ruby-pg/jobs/625409241

**Comment by eregon** [comment]
This is fixed with 9a22abd0844caae6ce1329dd39c08707235c96ba and 1bf2f6ac641454dae8498adb49bae15dc6b804db, by always returning a native pointer for `RSTRING_PTR`. It will be in the 20.0 release. Thank you for the bug report!

**Comment by eregon** [comment]
Ref: https://github.com/ged/ruby-pg/pull/331
